repository/src/main/java/org/apache/atlas/repository/store/graph/v2/EntityGraphMapper.java

```diff
@@ -2191,6 +2211,67 @@ private void addEdgesToContext(String guid, List<Object> newElementsCreated, Lis
         }
     }
 
+    private void addInternalProductAttr(AttributeMutationContext ctx, List<Object> createdElements, List<AtlasEdge> deletedElements, List<Object> currentElements) throws AtlasBaseException {
+        MetricRecorder metricRecorder = RequestContext.get().startMetricRecord("addInternalProductAttrForAppend");
+        AtlasVertex toVertex = ctx.getReferringVertex();
+        String toVertexType = getTypeName(toVertex);
+
+        if(currentElements.isEmpty() && createdElements.isEmpty() && deletedElements.isEmpty()){
+            RequestContext.get().endMetricRecord(metricRecorder);
+            return;
+        }
+
+        if (TYPE_PRODUCT.equals(toVertexType)) {
+
+            if(currentElements == null){
+                AtlasAttribute attribute           = ctx.getAttribute();
+                AtlasArrayType arrType             = (AtlasArrayType) attribute.getAttributeType();
+                AtlasType      elementType         = arrType.getElementType();
+                boolean        isStructType        = (TypeCategory.STRUCT == elementType.getTypeCategory()) ||
+                        (TypeCategory.STRUCT == attribute.getDefinedInType().getTypeCategory());
+
+                currentElements = (List) getCollectionElementsUsingRelationship(ctx.getReferringVertex(), attribute, isStructType);
+            }
+
+            if(ctx.getAttribute().getRelationshipEdgeLabel().equals(OUTPUT_PORT_PRODUCT_EDGE_LABEL)){
+                addOrRemoveInternalAttr(toVertex, OUTPUT_PORT_GUIDS_ATTR, createdElements, currentElements, deletedElements);
+            }
+            if (ctx.getAttribute().getRelationshipEdgeLabel().equals(INPUT_PORT_PRODUCT_EDGE_LABEL)) {
+                addOrRemoveInternalAttr(toVertex, INPUT_PORT_GUIDS_ATTR, createdElements, currentElements, deletedElements);
+            }
+        }else{
+           throw new AtlasBaseException(AtlasErrorCode.BAD_REQUEST, "Can not update product relations while updating any asset");
+        }
+        RequestContext.get().endMetricRecord(metricRecorder);
+    }
+
+    private void addOrRemoveInternalAttr(AtlasVertex toVertex, String internalAttr, List<Object> createdElements, List<Object> currentElements, List<AtlasEdge> deletedElements) {
+        List<String> portGuids = toVertex.getMultiValuedProperty(internalAttr, String.class);
+
+        if (CollectionUtils.isNotEmpty(currentElements) && CollectionUtils.isEmpty(portGuids)) {
+            List<String> currentGuids = currentElements.stream()
```

Comment:
found an issue
Case:
Assume a Daap have a couple of output ports with attributes having couple of guids
Update this Daap to remove both output ports, now attributes having empty list
Update this Daap to again add an output port among above 2, now attributes have couple guids including both previous guids
Expected: attributes should have only a guid which was added in last update

Reply:
Add a edge state ACTIVE check here so that deleted relations will not add up into currentGuids